checks/branch_protected.go

```diff
@@ -31,47 +33,88 @@ func init() {
 	registerCheck(CheckBranchProtection, BranchProtection)
 }
 
+type repositories interface {
+	Get(context.Context, string, string) (*github.Repository,
+		*github.Response, error)
+	ListReleases(ctx context.Context, owner string, repo string, opts *github.ListOptions) (
+		[]*github.RepositoryRelease, *github.Response, error)
+	GetBranchProtection(context.Context, string, string, string) (
+		*github.Protection, *github.Response, error)
+}
+
+type logger func(s string, f ...interface{})
+
 func BranchProtection(c *checker.CheckRequest) checker.CheckResult {
-	repo, _, err := c.Client.Repositories.Get(c.Ctx, c.Owner, c.Repo)
+	// Checks branch protection on both release and development branch
+	return checkReleaseAndDevBranchProtection(c.Ctx, c.Client.Repositories, c.Logf, c.Owner, c.Repo)
+}
+
+func checkReleaseAndDevBranchProtection(ctx context.Context, r repositories, l logger, ownerStr,
+	repoStr string) checker.CheckResult {
+	// Get release branches
+	releases, _, err := r.ListReleases(ctx, ownerStr, repoStr, &github.ListOptions{})
+	if err != nil {
+		return checker.MakeRetryResult(CheckBranchProtection, err)
+	}
+
+	var checks []checker.CheckResult
+	for _, release := range releases {
+		if release.TargetCommitish != nil {
```

Comment:
do you know when TargetCommitish may be nil? Shall we fail if it happens?

Reply:
doesn't seem like it. actually another point i thought of is that some of these branches may have overlap (e.g. be the dev branch). it won't affect the result since the result is a multicheckand, but it will reduce calls to get the branch protection.

Reply:
actually i think it can be nil https://github.com/google/go-github/blob/b235769d16066e228821a3e0fae66305b49021d6/github/github-accessors.go#L13295

it says "Unused if the Git tag already exists."

Reply:
Do you know what the comment actually mean? I don't :-)

How about failing and returning a specific error (e.g. `ErrCommitishNil`), so we can catch it upstream, log it and investigate when/how often this happens in practice. We can look at the logs from the cron job.

@azeemshaikh38 would this be ok? In the dockerfile pinning [check](https://github.com/ossf/scorecard/blob/main/checks/frozen_deps.go#L33), we return  `ErrInvalidDockerfile` for this reason. I don't think we're currently checking if this happens, though. Azeem, do you know? Maybe now is the time to start adding keyword to monitor the logs. wdut?

Reply:
Yes, it should be easy to setup monitoring to figure out how many errors (grouped by error type) we are seeing. Also, always a big +1 to returning specific errors that can be logged upstream.

However, I personally don't prefer enabling BranchProtection check in the cron job right away (it is currently disabled). I don't have confidence on what affect this check will have on our tokens. So we could do some local testing and build some confidence before enabling it in the cron job.

Reply:
added a distinct error!